buildkit/internal/transform/transforms.go

```diff
@@ -60,66 +62,104 @@ type virtualEnv struct {
-				adjustedLine = -node.EndLine
-				inPreamble = false
-			case pyenvCommand:
-				return nil, nil, fmt.Errorf("%s cannot appear before the first FROM", pyenvCommand)
+
+tokens:
+	for node := range tokens {
+		preamble.AddChild(node, node.StartLine, node.EndLine)
+		switch strings.ToUpper(node.Value) {
+		case argCommand:
+			if err := transformer.processArg(node); err != nil {
+				return nil, 0, err
 			}
-			preamble.AddChild(node, node.StartLine, node.EndLine)
-			continue
+		case fromCommand:
+			// finish preamble when we encounter the first FROM
+
+			// Try to use the astro-runtime base (non-onbuild) image. If we have any error, "fail safe"
+			// and leave it unmodified.
+			_ = transformer.ensureValidBaseImage(node)
+			adjustedLine = -node.EndLine
+			break tokens
+		case pyenvCommand:
+			return nil, 0, fmt.Errorf("%s cannot appear before the first FROM", pyenvCommand)
 		}
+	}
+	return preamble, adjustedLine, nil
+}
+
+func (transformer *Transformer) parseMain(tokens chan *parser.Node, lineNoAdjustment int) (*parser.Node, error) {
+	transformedAst := &parser.Node{StartLine: -1}
+
+	for node := range tokens {
 		switch strings.ToUpper(node.Value) {
 		case pyenvCommand:
 			pyenvNodes, err := transformer.processPyenv(node)
 			if err != nil {
-				return nil, nil, err
+				return nil, err
 			}
 			lineOffset := 0
 			for _, n := range pyenvNodes.Children {
-				transformedAst.AddChild(n, n.StartLine+adjustedLine, n.EndLine+adjustedLine)
+				transformedAst.AddChild(n, n.StartLine+lineNoAdjustment, n.EndLine+lineNoAdjustment)
 				lineOffset += (n.EndLine - n.StartLine) + 1
 			}
-			adjustedLine += lineOffset
+			lineNoAdjustment += lineOffset
 		default:
 			lineDiff := node.EndLine - node.StartLine
-			transformedAst.AddChild(node, adjustedLine+1, lineDiff+adjustedLine+1)
+			transformedAst.AddChild(node, lineNoAdjustment+1, lineDiff+lineNoAdjustment+1)
 		}
 	}
-	return preamble, transformedAst, nil
+	return transformedAst, nil
+}
+
+func (transformer *Transformer) Transform(dockerFile []byte) (*parser.Node, *parser.Node, error) {
+	ast, err := dockerfile.Parse(dockerFile)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// tokenStream is an "iterator" over the AST nodes in the Dockerfile
+	tokenStream := make(chan *parser.Node)
+	go func(ctx context.Context) {
+	tokens:
+		for _, node := range ast.Children {
+			select {
+			case <-ctx.Done():
+				break tokens
+			case tokenStream <- node:
+				continue
+			}
+		}
+		close(tokenStream)
+	}(ctx)
+	defer cancel()
+
+	premable, adjustedLine, err := transformer.parsePreamble(tokenStream)
+	if err != nil {
+		return nil, nil, err
+	}
+	transformedAst, err := transformer.parseMain(tokenStream, adjustedLine)
+
+	return premable, transformedAst, err
 }
 
-func processArg(node *parser.Node, buildArgs map[string]string) error {
+func (r *Transformer) processArg(node *parser.Node) error {
```

Comment:
Go style guidelines say to use the same receiver var name in all functions. I think rename all to `t`?

Reply:
Yes to `t` over `r` -- I wasn't sure if everything should be on the Transformer or if some of the functions could/should stay at top level

Reply:
I would say use the same receiver everywhere and whether the function is exported implies it is top level? I may not have grasped what you mean there though.

Reply:
Yeah, I guess in python this one would be a "static" function -- and go doesn't have/need such a concept, so this is currently just a "helper function".

Can make it a struct method for consistency though

Reply:
Thinking a little bit more about it -- it seems silly to make `validateVirtualEnv` and `extractPythonMajorMinor` have the receiver too

Reply:
Oh. Yeah I see your point. I missed changing one, I was thinking this diff comment was a suggestion!